### PR TITLE
TensorWithSeqLen: multiple fixes

### PR DESCRIPTION
**Pull request by danthe3rd**
* Fix wrong sanity check for LSE that didn't apply with varying seqlen
* CUTLASS: Fixed LSE shape when used with varying seqlen
* Add test for backward with `TensorWithSeqLen`
* Fix gradient flow through `TensorWithSeqLen.from_tensor_list` (+ add test)
* `TensorWithSeqLen`: fix with __getitem__ operation (and others with input tensors that are nested within the arguments)

### CODE REVIEW

File: xformers/ops/fmha/tensor_with_seqlen.py [hunk 1]
```diff
@@ -82,20 +87,34 @@ def new_empty(self, *args, **kwargs):
...
+            ) -> "TensorWithSeqLen":
+                out = TensorWithSeqLen(tensor)
+                out.cu_seqlen_py = cu_seqlen_py
+                out.cu_seqlen = torch.tensor(
+                    cu_seqlen_py, dtype=torch.int32, device=tensor.device
+                )
+                out.max_seqlen = max_seqlen
+                return out
+
+            @classmethod
```
Comment (fmassa) [inline xformers/ops/fmha/tensor_with_seqlen.py:104]: `@staticmethod` please ? :-)

**Review by fmassa — APPROVED** [review]
Thanks!
